packages/cardpay-sdk/sdk/utils/conversions.ts

```diff
@@ -117,40 +124,66 @@ export async function getGasPricesInNativeWei(
-  let rateAdjusted = adjustRate(rate);
-  return new BN(rateAdjusted.numerator.toString()).div(new BN(rateAdjusted.denominator.toString()));
+  return tokenPairRate;
 }
 
-export async function getUsdcToTokenRate(provider: JsonRpcProvider, tokenAddress: string): Promise<FixedNumber> {
+export async function getUsdcToTokenRate(provider: JsonRpcProvider, tokenAddress: string): Promise<TokenPairRate> {
   let network = await networkName(provider);
-  let usdStableCoinToken = getAddressByNetwork('usdStableCoinToken', network);
+  let usdcAddress = getAddressByNetwork('usdStableCoinToken', network);
+  let usdc = new Contract(usdcAddress, ERC20ABI, provider);
+  let token = new Contract(tokenAddress, ERC20ABI, provider);
+  let tokenPairRate = {
+    tokenInAddress: usdcAddress,
+    tokenOutAddress: tokenAddress,
+    tokenInDecimals: await usdc.callStatic.decimals(),
+    tokenOutDecimals: await token.callStatic.decimals(),
+    rate: FixedNumber.from(1),
+  };
 
-  if (usdStableCoinToken.toLowerCase() === tokenAddress.toLowerCase()) {
-    return FixedNumber.from(1);
+  if (usdcAddress.toLowerCase() !== tokenAddress.toLowerCase()) {
+    tokenPairRate.rate = adjustRate(await getTokenPairRate(provider, usdcAddress, tokenAddress));
   }
 
-  let rate = await tokenPairRate(provider, usdStableCoinToken, tokenAddress);
-  let numerator = FixedNumber.from(rate.numerator.toString());
-  let denominator = FixedNumber.from(rate.denominator.toString());
-  let scalar = FixedNumber.from(rate.scalar.numerator.toString()).divUnsafe(
-    FixedNumber.from(rate.scalar.denominator.toString())
-  );
-  const rateAdjusted = numerator.divUnsafe(denominator).mulUnsafe(scalar);
-  return rateAdjusted;
+  return tokenPairRate;
 }
 
-export function applyRateToAmount(rate: FixedNumber, amount: BigNumber): BigNumber {
+// return an amount in the smallest units of output token
+// if invert false, amount is the value of tokenIn in the smallest units
+// else amount is the value of tokenOut in the smallest units
+export function applyRateToAmount(tokenRate: TokenPairRate, amount: BigNumber, invert?: boolean | null): BigNumber {
   if (amount.isZero()) {
     return BigNumber.from(0);
   }
-  const convertedAmount = FixedNumber.from(amount.toString()).mulUnsafe(rate);
-  return BigNumber.from(convertedAmount.round(0).toString().split('.')[0]);
+  let ten = BigNumber.from(10);
+  let tokenInSmallestUnits = ten.pow(tokenRate.tokenInDecimals);
+  let tokenOutSmallestUnits = ten.pow(tokenRate.tokenOutDecimals);
+  if (invert) {
+    return amount
+      .mul(tokenInSmallestUnits)
+      .div(BigNumber.from(tokenRate.rate.mulUnsafe(FixedNumber.from(tokenOutSmallestUnits)).toString().split('.')[0]));
+  }
+  return BigNumber.from(
+    tokenRate.rate
+      .mulUnsafe(FixedNumber.from(amount))
+      .mulUnsafe(FixedNumber.from(tokenOutSmallestUnits))
+      .toString()
+      .split('.')[0]
+  ).div(tokenInSmallestUnits);
```

Comment:
Would be great to have a test for this function, also with tokens with different decimals. It looks very testable without the need for mocking etc